File: challenge-285/ppentchev/python/src/perl_weekly_285/defs.py
# SPDX-FileCopyrightText: Peter Pentchev <roam@ringlet.net>
# SPDX-License-Identifier: BSD-2-Clause
"""Common definitions for the perl-weekly-285 library."""

from __future__ import annotations

import dataclasses
import typing


if typing.TYPE_CHECKING:
    from typing import Final


VERSION: Final = "0.1.0"
"""The perl-weekly-285 library version, semver-like."""


@dataclasses.dataclass
class Error(Exception):
    """An error that occurred during the processing."""

    def __str__(self) -> str:
        """Provide a human-readable error message."""
        return f"{type(self).__str__} must be overridden for {self!r}"


@dataclasses.dataclass
class NoSolutionError(Error):
    """No solution could be found."""

    def __str__(self) -> str:
        """Provide a human-readable error message."""
        return f"Could not find a solution: {self!r}"

File: challenge-285/ppentchev/rust/src/defs.rs
// SPDX-FileCopyrightText: Peter Pentchev <roam@ringlet.net>
// SPDX-License-Identifier: BSD-2-Clause
//! Common definitions for the `perl-weekly` crate.

use anyhow::Error as AnyError;
use thiserror::Error;

/// An error that occurred during the doing of things.
#[derive(Debug, Error)]
#[non_exhaustive]
#[allow(clippy::error_impl_error)]
pub enum Error {
    /// Something went really, really wrong.
    #[error("fnmatch-regex internal error")]
    InternalError(#[source] AnyError),

    /// Either no solution or too many solutions or something else.
    #[error("No solution found: {0}")]
    NoSolution(String),

    /// Some known missing functionality.
    #[error("Not implemented yet: {0}")]
    NotImplemented(String),
}
